### PR TITLE
Fix handling of complex constant expression vectors inside metadata

### PR DESCRIPTION
**Pull request by vmaksimo**
Removed restriction for all the vector elements to be constant expressions as some of them might be `undef`.

### CODE REVIEW

**Comment by vmaksimo** [comment]
Tagging @MrSidims @AlexeySotkin to take a look

File: test/vector-metadata-constexpr.ll [hunk 1]
```diff
@@ -1,15 +1,25 @@
 ; RUN: llvm-as %s -o %t.bc
-; RUN: llvm-spirv -s %t.bc -o - | llvm-dis -o - | FileCheck %s
-; RUN: llvm-spirv %t.bc
+; RUN: llvm-spirv -s --spirv-ext=+SPV_INTEL_arbitrary_precision_integers,+SPV_INTEL_vector_compute %t.bc -o - | llvm-dis -o - | FileCheck %s
+; RUN: llvm-spirv --spirv-ext=+SPV_INTEL_arbitrary_precision_integers,+SPV_INTEL_vector_compute %t.bc
```
Comment (AlexeySotkin) [inline test/vector-metadata-constexpr.ll:3]: Why did you add the extensions?

Comment (vmaksimo) [reply test/vector-metadata-constexpr.ll:3]: I've added test case with the vector of size 1 which is supported only with the `SPV_INTEL_vector_compute` extension, and the second extension is not needed at all.
I'll change the test a bit to avoid using extensions.

Comment (AlexeySotkin) [reply test/vector-metadata-constexpr.ll]: Thanks!

Comment (vmaksimo) [reply test/vector-metadata-constexpr.ll]: done in 1bc8f8b

File: lib/SPIRV/SPIRVLowerConstExpr.cpp [hunk 2]
```diff
@@ -170,7 +170,7 @@ void SPIRVLowerConstExprBase::visit(Module *M) {
       auto LowerConstantVec = [&II, &LowerOp, &WorkList,
                                &M](ConstantVector *Vec,
                                    unsigned NumOfOp) -> Value * {
-        if (std::all_of(Vec->op_begin(), Vec->op_end(), [](Value *V) {
+        if (std::any_of(Vec->op_begin(), Vec->op_end(), [](Value *V) {
```
Comment (AlexeySotkin) [inline lib/SPIRV/SPIRVLowerConstExpr.cpp:173]: Do we really need this `if`? Is it possible to have a non-constant value in `ConstantVector`?

**Comment by AlexeySotkin** [comment]
@Fznamznon could you please take a look?

**Comment by AlexeySotkin** [comment]
In general this patch looks good to me.

However the whole approach with replacing constant expressions with non-constant instructions seems to be not quite efficient from size of generated module perspective. If we have multiple `llvm.dbg.value` with such complex constant expressions, we may increase size of the code significantly. Translating constant expressions to `OpSpecConstantOp` seems to be more efficient but perhaps harder solution.

File: lib/SPIRV/SPIRVLowerConstExpr.cpp (hunk 2 above)
Comment (Fznamznon) [reply lib/SPIRV/SPIRVLowerConstExpr.cpp:173]: From commit message and the test it seems `undef` if not a constant value, and can be met inside a `ConstantVector`.

**Review by Fznamznon — APPROVED** [review]
LGTM

File: lib/SPIRV/SPIRVLowerConstExpr.cpp (hunk 2 above)
Comment (Fznamznon) [reply lib/SPIRV/SPIRVLowerConstExpr.cpp:173]: I don't know what is the purpose of this `if` though.

**Comment by vmaksimo** [comment]
> In general this patch looks good to me.
> 
> However the whole approach with replacing constant expressions with non-constant instructions seems to be not quite efficient from size of generated module perspective. If we have multiple `llvm.dbg.value` with such complex constant expressions, we may increase size of the code significantly. Translating constant expressions to `OpSpecConstantOp` seems to be more efficient but perhaps harder solution.

If we choose to go with the more efficient approach (less size of a generated module) then we'll need to drop the existing lowering of constant expressions.
I have a draft patch that enables such an approach, but it does not rework the existing lowering pass. @AlexeySotkin does it seems a good idea to just drop the existing pass?

**Comment by vmaksimo** [comment]
>  UPD: I believe it'll be enough to drop the lowering of constant expression vectors, not all the pass. And this functionality will be covered with the translation to `OpSpecConstantOp`.

Per my experiments done, the approach from #1335 is not perfect and can't be finalized before the SPIR-V spec change is introduced (mentioned here: https://github.com/KhronosGroup/SPIRV-Tools/issues/4272).

I suggest proceeding with the solution offered by this PR.
